cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContextWithTX.java

```diff
@@ -0,0 +1,184 @@
+
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.batch;
+
+import co.cask.cdap.api.data.DatasetInstantiationException;
+import co.cask.cdap.api.data.batch.Split;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.mapreduce.AbstractMapReduce;
+import co.cask.cdap.api.mapreduce.MapReduceContext;
+import co.cask.cdap.api.mapreduce.MapReduceSpecification;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.data.Namespace;
+import co.cask.cdap.data2.datafabric.DefaultDatasetNamespace;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DynamicDatasetContext;
+import co.cask.cdap.data2.dataset2.NamespacedDatasetFramework;
+import co.cask.tephra.TransactionContext;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+/**
+ * Mapreduce job runtime context for use in the {@link AbstractMapReduce#beforeSubmit(MapReduceContext)} and
+ * {@link AbstractMapReduce#onFinish(boolean, MapReduceContext)} methods, which allow getting datasets
+ * that were not specified in the application specification. This allows users to get datasets at runtime
+ * that they didn't know about at compile time.
+ */
+public class BasicMapReduceContextWithTX implements MapReduceContext {
```

Comment:
Can you have it extends from `DynamicDatasetContext` instead? (that class get merged today through the notification merge).